contracts/IdoPairContract/src/IdoPairContract.cs

```diff
@@ -34,27 +37,45 @@ public static void _deploy(object data, bool update)
 
         public static void OnNEP17Payment(UInt160 from, BigInteger amount, object data)
         {
-            if (!IfCallFromIdoContractSwap()) throw new Exception("Not allowed call!");
-
-            if (GetAssetHash() == Runtime.CallingScriptHash)
+            if ((BigInteger)Storage.Get(Storage.CurrentContext, swapReceiveKey) == 1 && GetAssetHash() == Runtime.CallingScriptHash)
             {
                 SafeTransfer(GetTokenHash(), Runtime.ExecutingScriptHash, from, amount / Price);
             }
-        }
+            else if ((BigInteger)Storage.Get(Storage.CurrentContext, registerReceiveKey) == 1 && GetTokenHash() == Runtime.CallingScriptHash)
+            {
 
-        public static bool IfCallFromIdoContractSwap()
+            }
+            else 
+            {
+                ExecutionEngine.Abort();
+            }
+            ResetReceiveOnSwap();
+            ResetReceiveOnProjectRegister();
+        }
+        public static void SetReceiveOnContractRegister()
         {
-            Notification[] notifications = Runtime.GetNotifications(GetIdoContract());
+            if (Runtime.CallingScriptHash == GetIdoContract())
+            {
+                Storage.Put(Storage.CurrentContext, registerReceiveKey, 1);
+            }
+        }
 
-            foreach (var notification in notifications)
+        public static void SetReceiveOnSwap()
+        {
+            if (Runtime.CallingScriptHash == GetIdoContract())
             {
-                if (notification.EventName == "SwapAsset")
-                {
-                    return true;
-                }
+                Storage.Put(Storage.CurrentContext, registerReceiveKey, 1);
             }
+        }
 
-            return false;
+        private static void ResetReceiveOnProjectRegister()
+        {
+            Storage.Delete(Storage.CurrentContext, registerReceiveKey);
+        }
+
+        private static void ResetReceiveOnSwap()
+        {
+            Storage.Delete(Storage.CurrentContext, registerReceiveKey);
```

Comment:
What's the difference between these two methods?

Reply:
it's a mistake, ResetReceiveOnSwap should delete 'swapReceiveKey'

Reply:
Could you fix it and test it?